### PR TITLE
Update TypeVar typing on protobuf message.pyi

### PR DESCRIPTION
**Pull request by nipunn1313**
Update MergeFrom/CopyFrom to enforce that other matches self.

Add explicit bounded type vars for hasfield/clearfield/whichoneof, which mypy-protobuf can override. Better than using Any!

### CODE REVIEW

**Comment by benjaminp** [comment]
What is this for? I'm confused why a `Message` subclass would be generic over another `Message`. `MergeFrom` and `CopyFrom` always return the self-type, right? I also don't know why field names should be generic. Those are just strings, right?

**Comment by nipunn1313** [comment]
For Message:
MergeFrom/CopyFrom are supposed to take in another message of the same type. See Generic Self: https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self
With this change, it forces `self` and `other_msg` to be the same type.

For HasField/ClearField/WhichOneof:
mypy-protobuf autogenerates a `Literal["abc", "def"]` type for HasField/ClearField/WhichOneof which has been more precise than just a str
Each subclass has a different literal.

File: third_party/2and3/google/protobuf/message.pyi [hunk 1]
```diff
@@ -12,20 +12,23 @@ class _ExtensionDict:
...
+_OF = TypeVar("_OF", bound=str)  # oneof value
 
 if sys.version_info < (3,):
     _Serialized = Union[bytes, buffer, unicode]
 else:
     _Serialized = bytes
 
-class Message:
-    DESCRIPTOR: Any
+class Message(Generic[_M, _F, _OG, _OF]):
```
Comment (JelleZijlstra) [inline third_party/2and3/google/protobuf/message.pyi:25]: Doesn't look like this needs to be generic over `_M`

**Comment by benjaminp** [comment]
> For Message:
> MergeFrom/CopyFrom are supposed to take in another message of the same type. See Generic Self: https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self
> With this change, it forces `self` and `other_msg` to be the same type.

Yes, but why does `Message` need to be generic over that type variable?

> 
> For HasField/ClearField/WhichOneof:
> mypy-protobuf autogenerates a `Literal["abc", "def"]` type for HasField/ClearField/WhichOneof which has been more precise than just a str
> Each subclass has a different literal.

I guess the question is whether this is worth the cost of breaking all code using `Message` and `--disallow-any-generics`.

**Comment by nipunn1313** [comment]
Yeah - I realize this is a good point - it'll break anyone using disallow-any-generics, and create a pretty arcane way of having to deal with it (`Message[str,str,str,str]`). I don't think it's worth it for that.

Do you think there's a way to get best of both worlds here? I'd like to be able to inherit from `Message` while still being able to override HasField w/ more specific subtypes. One potential idea is to expose a `_Message` with generic types and have

```
class Message(_Message[str, str, str, str]):
    pass
```

for now, I will strip this diff down to just the `_M` which should be less controversial, and can separately work on a prototype of this other strategy to see if it could work.

**Comment by benjaminp** [comment]
current change lgtm now

Does it work to assign the literal types in the stubs and then `type: ignore` the overload errors?

**Comment by nipunn1313** [comment]
Interesting idea - can try it later. I'd be worried that it would `type: ignore` the entire definition (not just the overload error) but I can try it out

**Pull request merged**